File: GP.py
import numpy as np
import matplotlib.pyplot as plt

from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel, WhiteKernel

# 1. 真实霍尔测点坐标，单位可以是 mm
X_meas = np.array([
    [0, 0],
    [10, 0],
    [20, 0],
    [0, 10],
    [10, 10],
    [20, 10],
    [0, 20],
    [10, 20],
    [20, 20],
])

# 2. 霍尔芯片推出来的力，单位 N
F_meas = np.array([
    900.6642, 1900.1377, 1164.6233,
    1900.1377, 6258.707, 1900.1377,
    900.6642, 1900.1377, 900.6642])

# 3. 虚拟点网格
x = np.linspace(0, 20, 50)
y = np.linspace(0, 20, 50)
xx, yy = np.meshgrid(x, y)
X_virtual = np.column_stack([xx.ravel(), yy.ravel()])

# 4. GP kernel
kernel = (
    ConstantKernel(1.0, (1e-3, 1e3))
    * RBF(length_scale=30.0, length_scale_bounds=(1.0, 200.0))
    + WhiteKernel(noise_level=0.05, noise_level_bounds=(1e-6, 1.0))
)

# 5. 训练 GP
gp = GaussianProcessRegressor(
    kernel=kernel,
    normalize_y=True,
    n_restarts_optimizer=10,
    random_state=0
)

gp.fit(X_meas, F_meas)


# 6. 预测虚拟点力和不确定性
F_pred, F_std = gp.predict(X_virtual, return_std=True)

F_map = F_pred.reshape(xx.shape)
Std_map = F_std.reshape(xx.shape)


# 7. 画力分布
plt.figure()
plt.contourf(xx, yy, F_map, levels=30)
plt.scatter(X_meas[:, 0], X_meas[:, 1], c=F_meas, edgecolors="k")
plt.colorbar(label="Predicted force / N")
plt.xlabel("x / mm")
plt.ylabel("y / mm")
plt.axis("equal")
plt.title("GP reconstructed force field")
plt.show()

"""
# 8. 画不确定性
plt.figure()
plt.contourf(xx, yy, Std_map, levels=30)
plt.scatter(X_meas[:, 0], X_meas[:, 1], c=F_meas, edgecolors="k")
plt.colorbar(label="Uncertainty / N")
plt.xlabel("x / mm")
plt.ylabel("y / mm")
plt.axis("equal")
plt.title("GP uncertainty")
plt.show()
"""

print("Learned kernel:", gp.kernel_)

File: GP.m
clc;
clear;
close all;

%% 1. 真实霍尔测点坐标 (mm)
X_meas = [
    0, 0;
    10, 0;
    20, 0;
    0, 10;
    10, 10;
    20, 10;
    0, 20;
    10, 20;
    20, 20
];

%% 2. 霍尔芯片计算出的力 (N)
F_meas = [
    1085.53;
    2357.69;
    1085.53;
    2357.69;
    6258.71;
    2357.69;
    1085.53;
    2357.69;
    1085.53
];

%% 3. 创建连续预测网格，用于画图
x = linspace(0, 20, 50);
y = linspace(0, 20, 50);

[xx, yy] = meshgrid(x, y);
X_virtual = [xx(:), yy(:)];

%% 4. 建立 Gaussian Process Regression 模型
gprMdl = fitrgp( ...
    X_meas, ...
    F_meas, ...
    'KernelFunction', 'squaredexponential', ...
    'BasisFunction', 'constant', ...
    'Standardize', true, ...
    'Sigma', 0.05, ...
    'FitMethod', 'exact', ...
    'PredictMethod', 'exact');

%% 5. 预测连续网格力分布
[F_pred, F_std] = predict(gprMdl, X_virtual);

F_map = reshape(F_pred, size(xx));
Std_map = reshape(F_std, size(xx));

%% 6. 绘制力分布图
figure;

contourf(xx, yy, F_map, 30, 'LineColor', 'none');
hold on;

scatter( ...
    X_meas(:,1), ...
    X_meas(:,2), ...
    80, ...
    F_meas, ...
    'filled', ...
    'MarkerEdgeColor', 'k');

colorbar;
xlabel('x / mm');
ylabel('y / mm');
axis equal;
title('GP reconstructed force field');

%% 7. 输出所有整数坐标点的力数据

x_int = 0:1:20;
y_int = 0:1:20;

[xx_int, yy_int] = meshgrid(x_int, y_int);

X_int = [xx_int(:), yy_int(:)];

[F_int, F_int_std] = predict(gprMdl, X_int);

ResultTable = table( ...
    X_int(:,1), ...
    X_int(:,2), ...
    F_int, ...
    F_int_std, ...
    'VariableNames', { ...
        'x_mm', ...
        'y_mm', ...
        'PredictedForce_N', ...
        'Uncertainty_N'} ...
);

disp('整数坐标点力数据：');
disp(ResultTable);

%% 8. 保存整数坐标点力数据到 Excel
writetable(ResultTable, 'integer_coordinate_force_result.xlsx');

disp('整数坐标点力数据已保存为 integer_coordinate_force_result.xlsx');

%% 9. 输出学习后的模型参数
disp('Learned kernel parameters:');
disp(gprMdl.KernelInformation);

disp('Sigma noise:');
disp(gprMdl.Sigma);